### DISCUSSION TITLE
Avoidance Layers for Tilesets

### DISCUSSION THREAD

**Discussion by Advay17** [godotengine/godot-proposals#9412] [Engine Core]
It appears that the default editor does not have any way to paint avoidance layers like how it is possible to paint navigation or physics layers in a tileset. It makes npc pathing with tilesets difficult. Would it be possible to implement the painting of avoidance layers in tilesets to simplify pathing?

**Comment by AThousandShips** [comment]
How do you mean? Avoidance layers are set up on agents, not on maps/regions, there's an experimental feature on regions though, but it's experimental, and is specific to constrained obstacles *added* to the regions, and only to `NavigationRegion2D`, the `TileMap` doesn't utilize these regions directly, so it doesn't add those obstacles, so no avoidance layers apply

**Reply by smix8** [reply]
To add information to this. What the NavigationRegion2D `constrain_avoidance` property does is that it places a NavigationObstacle2D with vertices for each outline of the NavigationPolygon.

Something that users also can do in scripts, it is basically the same outlines, just that the winding order is important to set if the push agents out or in.

The reason why it is experimental is, this is a very naive way of constraining avoidance agents as without a margin between navigation mesh edges and avoidance edges it is likely that agents can get stuck between the edges.

That is also why this would not work at all for the TileMap. The TileMap already has near unsolvable issues with edge seams and the lack of collision margin with its internal navigation meshes. So using the TileMap tiled layouts for avoidance constrains would not only be a huge performance problem but also very low quality. It is the far better option to bake the TileMap with a NavigationRegion2D that also has the option available.

To make this work with the TileMap the TileMap would need to bake its outline edges from the tiles and combine them to larger units. The TileMap is not doing that for even more important features that have seam problems like physics or occlusion or navigation, so avoidance is even lower on its problems-list to solve.
